swiftlang/swift-docc
# add conformance availability data to overloaded declarations

Bug/issue #, if applicable: rdar://122589806

## Summary

Currently, the overloaded-declarations feature will save the overloaded symbols' declarations into the other symbols' Render JSON. This works well enough when symbols are overloaded by something that might appear their declarations, like function parameter types. However, overload groups may also be created for symbols with otherwise identical declarations, like with this example:

```swift
public class MyClass<T> {}

extension MyClass where T: Equatable {
    public func myFunc() {}
}

extension MyClass where T: Hashable {
    public func myFunc() {}
}
```

In situations like this, we need to also provide the conformance information in the overloaded declaration list, so that it may be rendered.

## Dependencies

A Swift-DocC-Render PR (not written yet) is required to actually display the conformance information.

## Testing

[This documentation catalog](https://github.com/user-attachments/files/16563803/asdf.docc.zip) contains a symbol graph generated from the above code. Although the associated renderer PR is not available yet, the Render JSON can still be inspected to verify the code is working as intended.

Steps:
1. `DOCC_JSON_PRETTYPRINT=YES swift run docc convert --enable-experimental-overloaded-symbol-presentation path/to/asdf.docc`
2. Open `asdf.docc/.docc-build/data/documentation/conformanceoverloads/myclass/myfunc()-6gquc.json`.
3. Inspect `primaryContentSections[0].declarations[0].otherDeclarations.declarations[0].conformance.constraints` (written as a `jq` filter) to ensure that it says "`T` conforms to `Equatable`." as Render JSON's inline-content data.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ n/a ] Updated documentation if necessary
